### DISCUSSION TITLE
Example how to call reqwest from Rocket

### DISCUSSION THREAD

**Discussion by alper** [rwf2/Rocket#2068] [Questions]
I found one example of how to do this but that seems to be outdated and does not compile: https://www.shawntabrizi.com/code/combining-rocket-with-reqwest-to-call-an-api-with-rust/

Does anybody have something current?

I have:
```
fn gmap() -> Result<String, Box<std::error::Error>> {

let body = reqwest::blocking::get(request_url)?.text()?;
```

but that also does not compile. (A common thread.)

**Comment by alper** [comment]
I found this that looks more current but also horribly contrived:
https://stuarth.github.io/rocket-error-handling/

Is that really the way to do this?

**Reply by alper** [reply]
I also can't implement Responder for reqwest::Error so what do I do?

**Reply by somehowchris** [reply]
You implement Responder for a wrapper of reqwest::Error or return a custom error on error.

I don't know your case but if you really need to have rocket as a server you are probably doing some business logic, therefore you shouldn't just plain forward the error you got from reqwest

**Reply by alper** [reply]
This is what the documentation says:

> If the error type E does not implement Responder, then the error is simply logged to the console, using its Debug implementation, and a 500 error is returned to the client.

So that doesn't really tally.

Also for this project I'd like to learn Rocket and get something to work. How the errors are actually handled is something of later concern. For now a don't care attitude seems more productive.

The example in the blog post works but I don't understand how a beginner could ever be expected to write something with a signature like: 
```
impl<'r, 'o: 'r> Responder<'r, 'o> for Error {
```

**Reply by somehowchris** [reply]
Have you ever written rust projects using implementations? It’s just an implementation with two lifetimes. You could also write it like that https://rocket.rs/v0.5-rc/guide/responses/#implementations

**Reply by blackghost1987** [reply]
I'm not sure the documentation is correct in that regard, a Result type won't work as a response if it doesn't implement Responder AFAIK. Maybe this has changed in 0.5 and the docs haven't been updated yet?

> Also for this project I'd like to learn Rocket and get something to work. How the errors are actually handled is something of later concern. For now a don't care attitude seems more productive.

If you REALLY don't care about the Errors, then the easiest way is to just use `unwrap()` or `expect()` to get rid of them, and not returning a Result type. This way your code will panic in case of an error, and Rocket will log the error and return an 500 Internal Server Error response automatically.

This is not the correct way to do things, you should not panic in production code (or only panic in really rare and unrecoverable situations), but it is a quick and dirty solution to your problem.

**Reply by blackghost1987** [reply]
Indeed this looks like a change in 0.5-rc1, just look at the following two pages:
https://api.rocket.rs/v0.4/rocket/response/trait.Responder.html
vs
https://api.rocket.rs/master/rocket/response/trait.Responder.html

0.4 has both
```
impl<'r, R: Responder<'r>, E: Debug> Responder<'r> for Result<R, E>
AND
impl<'r, R: Responder<'r>, E: Responder<'r> + Debug> Responder<'r> for Result<R, E>
```
so implementing Debug on the Error type is enough.

while 0.5-rc1 only has
```
impl<'r, 'o: 'r, 't: 'o, 'e: 'o, T, E> Responder<'r, 'o> for Result<T, E>
where
    T: Responder<'r, 't>,
    E: Responder<'r, 'e>, 
```

This seems like an intentional change based on the deprecation warning in 0.4:
```
Response was a non-`Responder` `Err`
This `Responder` implementation has been deprecated.
In Rocket v0.5, `Result<T, E>` implements `Responder` only if `E` implements `Responder`.
For the previous behavior, use `Result<T, Debug<E>>` where `Debug` is `rocket::response::Debug`."
```

The last line is your solution, much better than unwrapping. And of course the documentation will need to be fixed before releasing 0.5.

**Comment by mslot** [comment]
Use the async version works:

```rust
#[get("/sparks")]
async fn sparks() -> Json<Vec<Spark>> {
    let response = reqwest::get("https://www.google.com").await.unwrap().text().await.unwrap();
//...
}

```
I got the same error as you when i used the blocking feature.

**Reply by Sinope-Nanto** [reply]
I have the same problem as you, have you fixed it? I want to get my area from ifconfig.net/country-iso by using reqwest::blocking::get, but 
there will be an error : "Cannot drop a runtime in a context where blocking is not allowed. This happens when a runtime is dropped from within an asynchronous context." during compilation. Now, I use Command to request ifconfig.net/country-iso, it so stupid.

**Reply by Sinope-Nanto** [reply]
I don't want to use async version.
